Use separate figure instead of switching pyplot backend

jyu00 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Instead of switching the `pyplot` backend to a non-GUI (non-interactive) "Agg", this PR creates a separate `Figure`, attaches an `Agg` canvas, and use its axes for plotting. 


### Details and comments

Analysis/plotting is done in a separate thread (so it doesn't block the main thread), but `matplotlib` doesn't support GUI mode in a child thread. The previous solution was to switch to using a non-GUI backend (`Agg`) in the `requires_matplotlib` decorator, and switch it back once the plotting is done. This solution, however, as a couple of issues as documented in #352. 

After a bit more research, [this doc](https://matplotlib.org/stable/gallery/user_interfaces/canvasagg.html) shows how to use an agg backend directly, which is implemented in this PR. It replaces the `requires_matplotlib` decorator with the `get_non_gui_ax` function that returns an axes that can be used in a child thread.

mtreinish commented on qiskit_experiments/curve_analysis/curve_analysis.py:
```diff
@@ -21,7 +21,7 @@
 from typing import Any, Dict, List, Tuple, Callable, Union, Optional
 
-from matplotlib import pyplot
+from matplotlib import pyplot  # pylint: disable=unused-import
```
Is this import actually needed?

mtreinish reviewed (commented):
So this will definitely workaround issues where a child thread is trying to instantiate a gui backend by forcing figure/ax creation to be done with the agg backend explicitly. But, I'm wondering what if users want to use a different [non-gui backend](https://matplotlib.org/stable/tutorials/introductory/usage.html#the-builtin-backends). The big example I'm thinking of is the svg backend seems like it would be a useful choice here too. Sure the user can explicitly set the backend and pass ax to the class constructor, but I'm wondering if we want to give an option on the default backend.

jyu00 commented:
> But, I'm wondering what if users want to use a different [non-gui backend](https://matplotlib.org/stable/tutorials/introductory/usage.html#the-builtin-backends). The big example I'm thinking of is the svg backend seems like it would be a useful choice here too.

I originally picked `agg` because I read somewhere that it's the only built-in non-gui backend supported on all platforms. Of course I can't find the reference anymore. I switched to using `svg` as the default, since it works better in jupyter notebook, and all the tests still passed. 

I also added a global variable that allows the user to select a different backend. As much as I hate using a global variable, there is no existing per-session options (presumably user would want this applied to figures from all experiments). Using an environment variable is also a bit of a hassle as the user would need to supply the full module name so I can dynamically import. I'm open to suggestions though.

mtreinish commented:
> 
> I originally picked `agg` because I read somewhere that it's the only built-in non-gui backend supported on all platforms. Of course I can't find the reference anymore. I switched to using `svg` as the default, since it works better in jupyter notebook, and all the tests still passed.

It might have been that the `Agg` renderer is supported by all the user interfaces (except for the `macosx` one iirc). So there is an `Agg` variant for each GUI backend (like `TkAgg`). But for non-gui I **think** `Agg`, `PS`, `PDF`, and `SVG` are built in on all platforms without needing external libs (except for maybe pillow).

mtreinish commented on qiskit_experiments/matplotlib.py:
```diff
@@ -13,28 +13,28 @@
-        # The code below switches to a non-GUI backend "Agg" when creating the
-        # plot. An alternative is to run this in a separate process, but then
-        # we'd need to deal with pickling issues.
+    Analysis/plotting is done in a separate thread (so it doesn't block the
+    main thread), but matplotlib doesn't support GUI mode in a child thread.
+    This function creates a separate Figure and attaches a non-GUI
+    SVG canvas to it.
```
Maybe document the global variable usage here?

jyu00 replied:
This module is not actually included in the doc, so the docstring won't be easy to find. I ended up moving this `matplotlib.py` to `framework` and document the usage in its `__init__`.

mtreinish reviewed (commented):
This lgtm, except for there is no rendered documentation or release notes on how to change the default non-gui backend used in threads. I think svg is a sane default, but having an example that's included in the hosted documentation somewhere is something we definitely want to do.

mtreinish commented on qiskit_experiments/framework/__init__.py:
```diff
@@ -76,7 +76,19 @@
+Analysis/plotting is done in a separate child thread, so it doesn't block the
+main thread. Since matplotlib doesn't support GUI mode in a child threads, the
+figures generated during analysis need to use a non-GUI canvas. The default is
+`FigureCanvasSVG`, but you can change it to a different
```
```suggestion
:class:`~matplotlib.backends.backend_svg.FigureCanvasSVG`, but you can change it to a different
```

mtreinish commented on qiskit_experiments/framework/__init__.py:
```diff
@@ -76,7 +76,19 @@
+<https://matplotlib.org/stable/tutorials/introductory/usage.html#the-builtin-backends>`_
+by setting the ``qiskit_experiments.framework.matplotlib.default_figure_canvas``
+attribute. For example, you can set ``default_figure_canvas`` to
+`FigureCanvasAgg
+<https://matplotlib.org/stable/api/backend_agg_api.html>`_ to use the
```
```suggestion
:class:`~matplotlib.backends.backend_agg.FigureCanvasAgg` to use the
```

mtreinish commented on releasenotes/notes/matplotlib-backend-1e1033ce1815901f.yaml:
```diff
@@ -0,0 +1,17 @@
+    `non-interactive backend
+    <https://matplotlib.org/stable/tutorials/introductory/usage.html#the-builtin-backends>`_.
+    For example, you can set ``default_figure_canvas`` to
+    `FigureCanvasAgg
+    <https://matplotlib.org/stable/api/backend_agg_api.html>`_ to use the
```
```suggestion
    :class:`~matplotlib.backends.backend_agg.FigureCanvasAgg` to use the
```

mtreinish reviewed (approved):
LGTM, thanks for the updates. Just some small nits inline because we have intersphinx configured for matplotlib (https://github.com/Qiskit/qiskit-experiments/blob/main/docs/conf.py#L185) we can refer to classes in the matplotlib documentation directly in the docs
